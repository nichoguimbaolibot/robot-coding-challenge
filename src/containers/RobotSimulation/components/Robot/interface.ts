export interface IRobotProps {
  xAxisPlacement: number;
  yAxisPlacement: number;
  robotDirection: number;
  isPosition: boolean;
}
