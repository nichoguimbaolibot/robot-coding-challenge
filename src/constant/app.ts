export const MAX_BOXES = 5;
