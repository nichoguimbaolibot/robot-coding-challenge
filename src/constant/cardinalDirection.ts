export const NORTH = 'North';
export const EAST = 'East';
export const SOUTH = 'South';
export const WEST = 'West';

export const CARDINAL_DIRECTION = [
  {
    id: 1,
    direction: 0,
    position: NORTH,
  },
  {
    id: 2,
    direction: 90,
    position: EAST,
  },
  {
    id: 3,
    direction: 180,
    position: SOUTH,
  },
  {
    id: 4,
    direction: 270,
    position: WEST,
  },
];

export interface ICardinalDirection {
  direction?: number;
  id?: number;
  position?: string;
}
