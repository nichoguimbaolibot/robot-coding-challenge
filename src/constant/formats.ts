export const DATE_FORMAT = 'LTS';
