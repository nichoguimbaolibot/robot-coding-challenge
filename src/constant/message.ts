const MESSAGES = {
  INVALID_POSITION: 'Buddy~ It was an invalid position.',
  MOVEMENT_NOT_ALLOWED: 'Hey!! I will fall off the table!!',
  ROBOT_NOT_FOUND: 'Im not in the table broooo~',
};

export default MESSAGES;
