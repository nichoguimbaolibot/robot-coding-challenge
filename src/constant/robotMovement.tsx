export const ROBOT_MOVEMENT = {
  DEFAULT_X_AXIS_POSITION: 49,
  DEFAULT_Y_AXIS_POSITION: 420,
  DEFAULT_X_AXIS_MOVEMENT: 96,
  DEFAULT_Y_AXIS_MOVEMENT: 103,
};
