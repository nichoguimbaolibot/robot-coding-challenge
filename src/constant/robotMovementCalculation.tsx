import { ROBOT_MOVEMENT } from 'constant/robotMovement';

export const robotCalculation = {
  getXAxisMovement: (xAxis: number): number => {
    return (
      xAxis * ROBOT_MOVEMENT.DEFAULT_X_AXIS_MOVEMENT +
      ROBOT_MOVEMENT.DEFAULT_X_AXIS_POSITION
    );
  },
  getYAxisMovement: (yAxis: number): number => {
    return (
      yAxis * ROBOT_MOVEMENT.DEFAULT_X_AXIS_MOVEMENT +
      ROBOT_MOVEMENT.DEFAULT_X_AXIS_POSITION
    );
  },
};
