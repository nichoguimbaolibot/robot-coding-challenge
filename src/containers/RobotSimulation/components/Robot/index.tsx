import React, { useEffect } from 'react';
import { IRobotProps } from './interface';
import { robotCalculation } from 'constant/robotMovementCalculation';

const Robot: React.FunctionComponent<IRobotProps> = ({
  xAxisPlacement,
  yAxisPlacement,
  robotDirection,
  isPosition,
}): JSX.Element => {
  useEffect(() => {
    const { style } = document.getElementById('robot');
    if (!isPosition) {
      style.display = 'none';
      return;
    }
    const xAxisCalculation = robotCalculation.getXAxisMovement(xAxisPlacement);
    const yAxisCalculation = robotCalculation.getYAxisMovement(yAxisPlacement);

    const robotMovementCoordinates = `translate(${xAxisCalculation}px,${yAxisCalculation}px) rotate(${robotDirection}deg)`;

    style.display = 'block';
    style.transform = robotMovementCoordinates;
  }, [xAxisPlacement, yAxisPlacement, robotDirection, isPosition]);

  return (
    <section data-testid="robot-test" className="robot-section" id="robot" />
  );
};

export default Robot;
