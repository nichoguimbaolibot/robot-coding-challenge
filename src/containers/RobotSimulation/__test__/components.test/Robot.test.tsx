import React from 'react';
import { render, cleanup } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import Robot from 'containers/RobotSimulation/components/Robot';
import { robotCalculation } from 'constant/robotMovementCalculation';

afterEach(cleanup);

describe('test ROBOT component', () => {
  const TEST_ID = 'robot-test';
  it('renders ROBOT component', () => {
    const xAxisPlacement = 0;
    const yAxisPlacement = 0;
    const robotDirection = 0;
    const isPosition = false;
    const { asFragment } = render(
      <Robot
        xAxisPlacement={xAxisPlacement}
        yAxisPlacement={yAxisPlacement}
        robotDirection={robotDirection}
        isPosition={isPosition}
      />,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('expect ROBOT to be display BLOCK', () => {
    const xAxisPlacement = 0;
    const yAxisPlacement = 0;
    const robotDirection = 0;
    const isPosition = true;
    const { getByTestId } = render(
      <Robot
        xAxisPlacement={xAxisPlacement}
        yAxisPlacement={yAxisPlacement}
        robotDirection={robotDirection}
        isPosition={isPosition}
      />,
    );
    expect(getByTestId(TEST_ID).style.display).toBe('block');
  });

  it('expect ROBOT to be display NONE', () => {
    const xAxisPlacement = 0;
    const yAxisPlacement = 0;
    const robotDirection = 0;
    const isPosition = false;
    const { getByTestId } = render(
      <Robot
        xAxisPlacement={xAxisPlacement}
        yAxisPlacement={yAxisPlacement}
        robotDirection={robotDirection}
        isPosition={isPosition}
      />,
    );
    expect(getByTestId(TEST_ID).style.display).toBe('none');
  });

  it('expect ROBOT to have RIGHT COMPUTATION', () => {
    const xAxisPlacement = 0;
    const yAxisPlacement = 0;
    const robotDirection = 0;
    const isPosition = true;
    const xAxisCalculation = robotCalculation.getXAxisMovement(xAxisPlacement);
    const yAxisCalculation = robotCalculation.getYAxisMovement(yAxisPlacement);
    const { getByTestId } = render(
      <Robot
        xAxisPlacement={xAxisPlacement}
        yAxisPlacement={yAxisPlacement}
        robotDirection={robotDirection}
        isPosition={isPosition}
      />,
    );

    const robotMovementCoordinates = `translate(${xAxisCalculation}px,${yAxisCalculation}px) rotate(${robotDirection}deg)`;

    expect(getByTestId(TEST_ID).style.transform).toBe(robotMovementCoordinates);
  });
});
