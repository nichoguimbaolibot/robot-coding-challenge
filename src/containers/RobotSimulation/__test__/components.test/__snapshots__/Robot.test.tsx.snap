// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test ROBOT component renders ROBOT component 1`] = `
<DocumentFragment>
  <section
    class="robot-section"
    data-testid="robot-test"
    id="robot"
    style="display: none;"
  />
</DocumentFragment>
`;
